### PR TITLE
route to site selection when going to /theme

### LINKED ISSUE #53110: Theme Showcase: Errors occur when activating a theme from directly accessing the showcase URL.
Hi there, 

(Amended the issue body for clarity)

Recently the behavior of "All my sites" menu in Calypso multisite ( example here https://snipboard.io/whcAf7.jpg ) has changed-  Pressing Themes requires to choose a site first. 
To check that: in My Sites press Statistics > Swith Site >  All My Sites. That "All My Sites" menu can be entered not only from Statistics, but also from Domains, Posts, Pages, Plugins menus, but not Themes (I have to choose a site first to continue). 

That is not a big issue itself, but there is one error that is related to that changed behavior:

### Steps to reproduce the error

1. Go to https://wordpress.com/blog/category/themes/
2. Click on the first theme there – “New Theme: Twenty Twenty One”, it opens https://wordpress.com/blog/2021/01/14/twenty-twenty-one/
3. Near the bottom of the post, click on “Learn more about TwentyTwentyOne”, it opens the theme’s landing page: https://wordpress.com/theme/twentytwentyone
4. Click on the button near the top, “Activate this design FREE”
5. I’m presented a pop-over asking me, “How would you like to use Twenty Twenty-One on your site?” – example: https://d.pr/i/pKfQwi
6. Click on “Activate Twenty Twenty-One” button to activate the theme.
**7. Immediately get an error message in the top-right that says, “Unable to activate theme. Contact support.” – example: https://d.pr/i/bQ8MNh**

### Solution
**Possible solution: after step 4 to show the site selection window like this https://snipboard.io/WlU4tc.jpg**

On WordPress.com desktop app same issue is already solved:  You can go to My Sites > Switch Site > All my sites > Themes> Twenty Twenty One > Activate this design FREE > **a popup appears to choose a site** > Activate > Activate Twenty Twenty-One >Edit homepage. 

Also, "Themes" menu item can be removed from "All My Sites" menu (https://snipboard.io/whcAf7.jpg), because it does not make any sense to show a menu item that does not work as expected: does not show Themes available for all websites.

#### Level of impact 
Potentially can block paid theme purchases for multisite users.

#### DETAILS
That behavior started 21/05/21, a day before it worked fine.
This is similar to [#52789](https://github.com/Automattic/wp-calypso/issues/52789).

### PR DESCRIPTION
**Pull request by andres-blanco**
#### Changes proposed in this Pull Request

This PR adds a site selection screen if the user is logged in and navigates to `/theme/SLUG`

#### Testing instructions

* Check that going to `/theme/twentytwentyone` show the site selector screen. Like this:
image

* When you select a site verify that you see the theme selected and the site is the one you selected `/theme/twentytwentyone/[YOUR_SITE]`
* Check that when you click the Activate button both previews looks ok in the popup.

I moved the `selectSiteIfLoggedIn` middleware to a common place so it can be reused from `/theme` and `/themes`. It would be nice to check that the themes showcase did not change its behaviour.
* Go to `/themes`
* Verify you see the site picker screen
* Select a site
* Verify you see the theme showcase of your selected site

Related to https://github.com/Automattic/wp-calypso/issues/53110

### CODE REVIEW

**Comment by blackjackkent** [comment]
Behavior on this looks good! 

The "Select a site to open **Theme**" header is a little generic; is there a way to have that more specifically reference the theme being navigated to?

**Comment by sixhours** [comment]
I haven't had a chance to review this yet, but I would suggest that we hold off on merging it until after the Theme Showcase sprint changes have launched and we've confirmed we haven't broken anything. 😅

**Comment by andres-blanco** [comment]
> Behavior on this looks good!
> 
> The "Select a site to open **Theme**" header is a little generic; is there a way to have that more specifically reference the theme being navigated to?

@blackjackkent I managed to put a better title including the theme name. Of course I accept suggestions for better wording :)

> I haven't had a chance to review this yet, but I would suggest that we hold off on merging it until after the Theme Showcase sprint changes have launched and we've confirmed we haven't broken anything. 😅

@sixhours don't worry! I will hold until **the great merge**

File: client/my-sites/theme/index.web.js
Comment (DanielSmithMichigan) [inline client/my-sites/theme/index.web.js]: How about:
Select a site to view **Spearhead**
or:
Select a site to see **Spearhead**

I think "details" at the end makes it just a tiny bit too wordy



This makes it feel closer to the Theme's page as well, in my opinion

**Comment by DanielSmithMichigan** [comment]
These code changes make sense and it tested out alright. I just had one suggestion for language.

https://user-images.githubusercontent.com/10274366/124799973-1e4acc80-df23-11eb-8310-717e68023b9b.mov

https://user-images.githubusercontent.com/10274366/124799992-230f8080-df23-11eb-9e7a-0f7ce4d422e7.mov

**Comment by mreishus** [comment]
Working for me in testing as well :+1:

File: client/my-sites/theme/index.web.js
Comment (andres-blanco) [reply client/my-sites/theme/index.web.js]: That's indeed better. I will fix it!

**Review by kwight — APPROVED** [review]

**Comment by kwight** [comment]
Nice, I think we're good here @andres-blanco ! Since there's SSR changes involved, just make sure to test well on staging and be ready to revert if we see anything weird. 🙃

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6180697

Thank you @andres-blanco for including a screenshot in the description! This is really helpful for our translators.

File: client/controller/index.web.js
Comment (tyxla) [inline client/controller/index.web.js:178]: Is there any chance this could not live in the primary controller, but in the themes one for example? It caused a huge (~ 20%) inflation on the main entry point (see http://iscalypsofastyet.com/).

cc @Automattic/team-calypso-frameworks @sgomes for awareness.

**Comment by sgomes** [comment]
Hey folks!

I should emphasise @tyxla's concern here. This PR appears to have caused a massive regression in Calypso entry point size which could negatively impact performance for most routes in Calypso.

I'll be taking a look at this PR to see if we can figure out a way of avoiding the massive inflation that it appears to have caused in bundle size.

In the future, I would urge any teams making changes to core elements such as the root controller to proactively reach out to @Automattic/team-calypso ahead of time, such as by asking for a review, as they'll be happy to help you make sure the changes follow best practices. Also feel free to reach out to me if you have any performance concerns at any point.

File: client/controller/index.web.js
Comment (jsnajdr) [reply client/controller/index.web.js:178]: Yes, site selection is specific for `my-sites` and should be at least in `my-sites/controller`, never in the top-level `calypso/controller`. Or it could be in `my-sites/themes/controller`.

File: client/layout/style.scss
Comment (jsnajdr) [inline client/layout/style.scss:47]: This style should also be somewhere in `my-sites/theme`. We should be removing section-specific styles from the common layout stylesheet instead of adding new ones.

**Comment by sgomes** [comment]
I attempted a fix for the entry point size increase in #54611, by moving the method to the `my-sites` controller. If you're following this thread, please take a look!

File: client/controller/index.web.js
Comment (andres-blanco) [reply client/controller/index.web.js:178]: @tyxla thanks for bringing it up! @sgomes already submitted a fix, I will be reviewing it

**Comment by tyxla** [comment]
Thanks @sgomes 🙌

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6180697).

**Pull request merged**
